stacks-core/src/codec.rs

```diff
@@ -0,0 +1,40 @@
+use std::io;
+
+use thiserror::Error;
+
+use crate::StacksResult;
+
+#[derive(Error, Debug)]
+pub enum CodecError {
+    #[error("Could not serialize or deserialize: {0}")]
+    IoError(#[from] io::Error),
+}
+
+pub type CodecResult<T> = Result<T, CodecError>;
+
+pub trait Codec {
+    fn codec_serialize<W: io::Write>(&self, dest: &mut W) -> io::Result<()>;
+    fn codec_deserialize<R: io::Read>(data: &mut R) -> io::Result<Self>
+    where
+        Self: Sized;
+
+    fn serialize<W: io::Write>(&self, dest: &mut W) -> StacksResult<()> {
+        self.codec_serialize(dest)
+            .map_err(|err| CodecError::IoError(err).into())
+    }
+
+    fn deserialize<R: io::Read>(data: &mut R) -> StacksResult<Self>
+    where
+        Self: Sized,
+    {
+        Self::codec_deserialize(data).map_err(|err| CodecError::IoError(err).into())
+    }
```

Comment:
I don't feel like these methods really add anything. I'd prefer to have a provided serialize function outside the trait for specific return types, similar to how the bitcoin crate does it https://docs.rs/bitcoin/0.30.1/bitcoin/consensus/encode/fn.serialize.html

Reply:
I'm not sure I understand. That `serialize()` function from the `bitcoin` crate requires the type to implement `Encodable` which is basically the same as `Codec` here.

I could add a serialize function to the `stacks-core` if you think that would improve ergonomics?

Reply:
Yeah I agree about the required functions in the Codec trait. I was referring to the two provided methods here which just seems to wrap the `codec_serialize` and `codec_deserialize` functions with a new error type.

Reply:
Discussed this on Slack and decided to leave potential updates for future PRs.